components/patient-data/impl/src/main/java/org/phenotips/data/internal/EmptyGenotypeObjectsRemover.java

```diff
@@ -0,0 +1,97 @@
+/*
+ * See the NOTICE file distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see http://www.gnu.org/licenses/
+ */
+package org.phenotips.data.internal;
+
+import org.phenotips.Constants;
+import org.phenotips.data.events.PatientChangingEvent;
+
+import org.xwiki.component.annotation.Component;
+import org.xwiki.model.EntityType;
+import org.xwiki.model.reference.EntityReference;
+import org.xwiki.observation.AbstractEventListener;
+import org.xwiki.observation.event.Event;
+
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+
+import javax.inject.Named;
+import javax.inject.Singleton;
+
+import org.apache.commons.lang3.StringUtils;
+
+import com.xpn.xwiki.doc.XWikiDocument;
+import com.xpn.xwiki.objects.BaseObject;
+import com.xpn.xwiki.objects.BaseStringProperty;
+
+/**
+ * Removes gene and variant objects from the document if key fields are empty, gene field for genes and cdna field for
+ * variants.
+ *
+ * @version $Id$
+ * @since 1.3M2
+ */
+@Component
+@Named("empty-genotype-objects-remover")
+@Singleton
+public class EmptyGenotypeObjectsRemover extends AbstractEventListener
```

Comment:
We will leave this improvement for later. Other comments addressed and PR rebased.